Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs

```diff
@@ -180,6 +180,106 @@ public AsyncFunctionAssertions(Func<TTask> subject, IExtractExceptions extractor
+            .BecauseOf(because, becauseArgs)
+            .FailWith("Expected {context} to throw {0} within {1}{reason}, but found <null>.",
+                typeof(TException), timeSpan);
+
+        if (success)
+        {
+            Exception caughtException = await InvokeWithInterceptionAsync(timeSpan);
+            return AssertThrows<TException>(caughtException, timeSpan, because, becauseArgs);
+        }
+
+        return new ExceptionAssertions<TException>(Array.Empty<TException>());
+    }
+
+    private ExceptionAssertions<TException> AssertThrows<TException>(
+        Exception exception, TimeSpan timeSpan, string because, object[] becauseArgs)
+        where TException : Exception
+    {
+        TException[] expectedExceptions = Extractor.OfType<TException>(exception).ToArray();
+
+        Execute.Assertion
+            .BecauseOf(because, becauseArgs)
+            .WithExpectation("Expected a <{0}> to be thrown within {1}{reason}, ",
+                typeof(TException), timeSpan)
+            .ForCondition(exception is not null)
+            .FailWith("but no exception was thrown.")
+            .Then
+            .ForCondition(expectedExceptions.Any())
+            .FailWith("but found <{0}>: {1}{2}.",
+                exception?.GetType(),
+                Environment.NewLine,
+                exception)
+            .Then
+            .ClearExpectation();
+
+        return new ExceptionAssertions<TException>(expectedExceptions);
+    }
+
+    private async Task<Exception> InvokeWithInterceptionAsync(TimeSpan timeout)
+    {
+        try
+        {
+            // For the duration of this nested invocation, configure CallerIdentifier
+            // to match the contents of the subject rather than our own call site.
+            //
+            //   Func<Task> action = async () => await subject.Should().BeSomething();
+            //   await action.Should().ThrowAsync<Exception>();
+            //
+            // If an assertion failure occurs, we want the message to talk about "subject"
+            // not "await action".
+            using (CallerIdentifier.OnlyOneFluentAssertionScopeOnCallStack()
+                       ? CallerIdentifier.OverrideStackSearchUsingCurrentScope()
+                       : default)
+            {
+                (TTask result, TimeSpan remainingTime) = InvokeWithTimer(timeout);
+                if (remainingTime < TimeSpan.Zero)
+                {
+                    // timeout reached without exception
+                    return null;
+                }
+
+                if (result.IsFaulted)
+                {
+                    // exception in synchronous portion
+                    return result.Exception!.GetBaseException();
+                }
+
+                if (await CompletesWithinTimeoutAsync(result, remainingTime))
+                {
+                    // completed without exception
+                    return null;
+                }
```

Comment:
I think we can write it like this to all keep lines covered.
```cs
_ = await CompletesWithinTimeoutAsync(result, remainingTime);

// completed without exception
return null;
```

In the other usages `CompletesWithinTimeoutAsync(` we the Boolean result in the `ForCondition`, whereas here we use the returned exception from this method.

Reply:
You are right, but it's a bit hard explain, WHY we can ignore the result, at least while "quick reading the code"...
So, I tried to write an explanation comment.
Thanks!